File: 2025/python/day11.py
from utils import get_line_data, timeit, memoize


GRAPH = {}


def build_graph(data):
    for line in data:
        device, outputs = line.split(": ", 1)
        GRAPH[device.strip()] = [out.strip() for out in outputs.split()]

    GRAPH["out"] = []


@memoize
def count_paths(curr, end, has_visited_dac, has_visited_fft):
    if curr == "fft":
        has_visited_dac = True
    if curr == "dac":
        has_visited_fft = True

    if curr == end:
        return 1 if has_visited_dac and has_visited_fft else 0

    return sum(
        count_paths(neighbor, end, has_visited_dac, has_visited_fft)
        for neighbor in GRAPH[curr]
    )


@timeit
def main():
    data = get_line_data("11")
    build_graph(data)

    part_one = count_paths("you", "out", True, True)
    print(part_one)

    part_two = count_paths("svr", "out", False, False)
    print(part_two)

File: 2025/rust/day11.rs
use crate::y2025::utils;
use std::collections::HashMap;

type Graph = HashMap<String, Vec<String>>;

fn build_graph(data: &[String]) -> Graph {
    let mut graph = Graph::new();

    for line in data {
        if let Some((device, outputs)) = line.split_once(": ") {
            let device = device.trim().to_string();
            let outputs: Vec<String> = outputs
                .split_whitespace()
                .map(|s| s.trim().to_string())
                .collect();
            graph.insert(device, outputs);
        }
    }

    graph.entry("out".to_string()).or_insert_with(Vec::new);

    graph
}

fn count_paths(
    graph: &Graph,
    cache: &mut HashMap<(String, String), i64>,
    start: &str,
    end: &str,
) -> i64 {
    let key = (start.to_string(), end.to_string());

    if let Some(&cached) = cache.get(&key) {
        return cached;
    }

    if start == end {
        cache.insert(key, 1);
        return 1;
    }

    if start == "out" {
        cache.insert(key, 0);
        return 0;
    }

    let result = graph
        .get(start)
        .map(|neighbors| {
            neighbors
                .iter()
                .map(|neighbor| count_paths(graph, cache, neighbor, end))
                .sum()
        })
        .unwrap_or(0);

    cache.insert(key, result);
    result
}

pub fn main() {
    let data = utils::get_line_data("11");
    let graph = build_graph(&data);

    let mut cache = HashMap::new();

    let part_one = count_paths(&graph, &mut cache, "you", "out");
    println!("{}", part_one);

    let part_two = count_paths(&graph, &mut cache, "svr", "fft")
        * count_paths(&graph, &mut cache, "fft", "dac")
        * count_paths(&graph, &mut cache, "dac", "out")
        + count_paths(&graph, &mut cache, "svr", "dac")
            * count_paths(&graph, &mut cache, "dac", "fft")
            * count_paths(&graph, &mut cache, "fft", "out");
    println!("{}", part_two);
}
